make_abi_library.py

```diff
@@ -15,34 +12,41 @@
 STATS = {}
 
 
-def has_function(args):
+def has_function(seen=None, *args):
   """ Check if any member of `args` is a function, or will lead to a function
       For example, a typedef, a use or a pointer with many indirections
   """
   global FUNCTION_CACHE
 
+  # maintain a cache of types visited in this has_function stack
+  if seen is None:
+    seen = set()
+
   for arg in args:
-    # Check if this specific argument has already been processed
+    # Check if this argument has already been processed
+    # if yes, exit to avoid infinite recursion
+    if arg in seen:
+      return False
+
+    # we have now seen this argument, to prevent infinite recursion
+    seen.add(arg)
+
+    # check if we have a cached determination
     cached = FUNCTION_CACHE.get(arg, None)
     if cached is not None:
       if cached == True:
```

Comment:
I think you want `return cached` when the `cached is not None` is `True`.

Reply:
So cached would only ever be `True` for now, but your version is easier to understand and less code; will change.